design/clean_artifacts_in_csi_flow.md

```diff
@@ -0,0 +1,422 @@
+	if err != nil {
+		p.log.Errorf("Fail to convert VS %s to unstructured", operationID)
+		return nil, errors.WithStack(err)
+	}
+
+	p.log.Infof(`Returning from VolumeSnapshotRestoreItemAction with 
+		no additionalItems`)
+
+	return &velero.RestoreItemActionExecuteOutput{
+		UpdatedItem:     &unstructured.Unstructured{Object: vsMap},
+		AdditionalItems: []velero.ResourceIdentifier{},
+		OperationID:     operationID,
+	}, nil
+}
+```
+
+Because VolumeSnapshotContent is not restored dynamically, if we run two restores in the same cluster, the second restore will fail due to the VolumeSnapshotContent is there in the cluster, and it is already bound to an existing VolumeSnapshot.
+
+To avoid this issue, it's better to delete the restored VolumeSnapshot and VolumeSnapshotContent after restore completes.
+
+``` go
+func (p *volumeSnapshotRestoreItemAction) Progress(
+	operationID string,
+	restore *velerov1api.Restore,
+) (velero.OperationProgress, error) {
+	progress := velero.OperationProgress{}
+
+	// The operationID format should be "vsNamespace/vsName".
+	if operationID == "" || !strings.Contains(operationID, "/") {
+		return progress, riav2.InvalidOperationIDError(operationID)
+	}
+
+	logger := p.log.WithFields(logrus.Fields{
+		"Action":      "VolumeSnapshotRestoreItemAction",
+		"OperationID": operationID,
+		"Namespace":   restore.Namespace,
+	})
+
+	vsc := new(snapshotv1api.VolumeSnapshotContent)
+	vs := new(snapshotv1api.VolumeSnapshot)
+
+	operationIDParts := strings.Split(operationID, "/")
+
+	if err := p.crClient.Get(
+		context.TODO(),
+		crclient.ObjectKey{Namespace: operationIDParts[0], Name: operationIDParts[1]},
+		vs,
+	); err != nil {
+		logger.Errorf("Fail to get the VolumeSnapshot %s", operationID)
+		return progress, err
+	}
+
+	if vs.Spec.Source.VolumeSnapshotContentName == nil {
+		logger.Errorf("VolumeSnapshot %s referenced VSC is nil", operationID)
+		return progress, errors.Errorf("VolumeSnapshot %s referenced VSC is nil", operationID)
+	}
+
+	if err := p.crClient.Get(
+		context.TODO(),
+		crclient.ObjectKey{
+			Name: *vs.Spec.Source.VolumeSnapshotContentName,
+		},
+		vsc,
+	); err != nil {
+		logger.Errorf("Fail to get the VolumeSnapshotContent %s",
+			*vs.Spec.Source.VolumeSnapshotContentName,
+		)
+		return progress, err
+	}
+
+	if (vsc.Status != nil && boolptr.IsSetToTrue(vsc.Status.ReadyToUse)) &&
+		(vs.Status != nil && boolptr.IsSetToTrue(vs.Status.ReadyToUse)) {
+		// VS and VSC are ready. Delete them to make sure they will not cause conflict for the next restore.
+		if err := p.crClient.Delete(context.TODO(), vs); err != nil {
+			logger.Warnf("Fail to delete VS %s after VS turns ready %s.", vs.Namespace+vs.Name, err.Error())
+		}
+		if err := p.crClient.Delete(context.TODO(), vsc); err != nil {
+			logger.Warnf("Fail to delete VSC %s after VSC turns ready %s.", vsc.Name, err.Error())
+		}
+	} else {
+		logger.Debugf("VolumeSnapshot or VolumeSnapshotContent are not ready yet.")
+		return progress, nil
```

Comment:
When the snapshot creation is in progress, it's better we deliver some info, like filling something `progress.Description`

Reply:
Thanks. Modified.